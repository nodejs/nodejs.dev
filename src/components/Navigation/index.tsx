import React from 'react';
import { FormattedMessage } from 'react-intl';
import classnames from 'classnames';
import styles from './index.module.scss';

interface Props {
  isOpen: boolean;
  label: string;
  toggleNavigation: () => void;
}

const NavigationContainer = ({
  label,
  isOpen,
  children,
  toggleNavigation,
}: React.PropsWithChildren<Props>) => {
  const navigationClasses = classnames(styles.navigation, {
    [styles.navigationFixed]: isOpen,
  });

  return (
    <nav aria-label={label} className={navigationClasses} id="main-navigation">
      <button
        type="button"
        className={styles.navigationOpen}
        onClick={toggleNavigation}
        aria-expanded={isOpen}
      >
        <FormattedMessage id="containers.navigation.title" />
      </button>
      {children}
    </nav>
  );
};

export default NavigationContainer;
