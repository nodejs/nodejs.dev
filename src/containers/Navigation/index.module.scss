.navigation,
%navigation {
  background-color: var(--color-fill-side-nav);
  box-sizing: border-box;
  clip-path: polygon(0 0, 100% 0, 100% 100%, 0 100%);
  height: calc(100vh - var(--nav-height));
  max-width: 36rem;
  overflow: auto;
  overflow-x: hidden;
  padding: 0 var(--space-20);
  position: sticky;
  top: var(--nav-height);
  width: 100%;
  z-index: 998;

  @media (max-width: 900px) {
    height: 42px;
    max-width: unset;
    overflow: hidden;
    position: relative;
    top: 0;
    transition: all 0.5s cubic-bezier(0.23, 1, 0.32, 1);
  }

  &__open,
  &%__open {
    background: var(--color-text-primary);
    background-clip: content-box;
    border: none;
    color: var(--color-text-primary);
    display: none;
    font-weight: var(--font-weight-light);
    height: calc(24px + 3px);
    line-height: 0;
    margin-left: -0.8rem;
    overflow: visible;
    padding: 12px 0;
    position: sticky;
    text-indent: 3.2rem;
    top: 0.8rem;
    transition: color 0.28s, background-color 0.28s;
    user-select: none;
    width: 20px;
    z-index: 9999;

    @media (max-width: 900px) {
      display: unset;
    }

    &::before {
      background-color: var(--color-text-primary);
      content: '';
      height: 3px;
      left: 0;
      position: absolute;
      top: calc(50% - 1.5px);
      transform: translateY(-250%);
      transition: transform 0.28s;
      width: 100%;
    }

    &::after {
      background-color: var(--color-text-primary);
      content: '';
      height: 3px;
      left: 0;
      position: absolute;
      top: calc(50% - 1.5px);
      transform: translateY(250%);
      transition: transform 0.28s;
      width: 100%;
    }
  }

  &--fixed,
  &%--fixed {
    @media (max-width: 900px) {
      height: 100vh;
      overflow-x: hidden;
      overflow-y: scroll;
      position: sticky;
      top: 0;
      z-index: 999;
    }
  }

  &--fixed &__open {
    background-color: transparent;
    color: transparent;

    &::before {
      transform: rotate(-45deg);
    }

    &::after {
      transform: rotate(45deg);
    }
  }
}
