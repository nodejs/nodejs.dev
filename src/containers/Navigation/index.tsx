import React, { useState, useRef, useEffect } from 'react';
import NavigationSection from '../../components/NavigationSection';
import { NavigationSectionData, NavigationSectionItem } from '../../types';
import { isMobileScreen } from '../../util/isScreenWithinWidth';
import { scrollTo, calcNavScrollParams } from '../../util/scrollTo';

interface Props {
  sections: NavigationSectionData;
  currentSlug: string;
  previousSlug: string;
  label: string;
}

const Navigation = ({
  sections,
  currentSlug,
  previousSlug,
  label,
}: Props): JSX.Element => {
  const [isOpen, setIsOpen] = useState<boolean>(false);
  const navElement = useRef<HTMLElement | null>(null);
  const toggle = (): void => setIsOpen(!isOpen);
  const onItemClick = (): void => {
    if (isMobileScreen()) {
      toggle();
    }
  };

  useEffect(() => {
    if (typeof window !== 'undefined') {
      const nav = navElement.current;
      if (nav) {
        //  Get current link element
        const currentLinkElem = document.getElementById(`link-${currentSlug}`);
        // Get scroll position for current link
        const { newScrollPos, scrollWindow, scrollTime } = calcNavScrollParams(
          currentLinkElem?.getBoundingClientRect().top || 0,
          nav
        );
        // If it's mobile screen and nav is not open, do nothing
        if (isMobileScreen() && !isOpen) return;
        // If it's mobile screen, directly scroll to current element when nav opens
        if (isMobileScreen()) {
          scrollTo(newScrollPos, scrollWindow, scrollTime);
        } else if (!previousSlug) {
          // If there's no previous slug, directly scroll to the current link
          nav.scrollTop = newScrollPos;
        } else {
          // Get previous link element
          const prevLinkElem = document.getElementById(`link-${previousSlug}`);
          // Get previous element's position
          const prevElemScroll = calcNavScrollParams(
            prevLinkElem?.getBoundingClientRect().top || 0,
            nav
          );
          // First scroll directly to previous element
          nav.scrollTop = prevElemScroll.newScrollPos;
          // Now scroll to current element with animation
          scrollTo(newScrollPos, scrollWindow, scrollTime);
        }
      }
    }
  }, [currentSlug, previousSlug, isOpen]);

  const className = isOpen ? 'side-nav side-nav--open' : 'side-nav';

  const readSections: Set<NavigationSectionItem['slug']> = new Set();
  // Assume section items up to the one currently open have been read. Track
  // their unique slugs in `readSections` set.
  Object.keys(sections).some((sectionKey): boolean => {
    let isCurrentSlug = false;
    sections[sectionKey].some((sectionItem): boolean => {
      isCurrentSlug = sectionItem.slug === currentSlug;
      if (!isCurrentSlug) {
        readSections.add(sectionItem.slug);
      }

      return isCurrentSlug;
    });

    return isCurrentSlug;
  });

  return (
    <nav aria-label={label} className={className} ref={navElement}>
      <button type="button" className="side-nav__open" onClick={toggle}>
        Menu
      </button>
      {Object.keys(sections).map(
        (sectionKey: string): false | JSX.Element =>
          sectionKey !== 'null' && (
            <NavigationSection
              key={sectionKey}
              title={sectionKey}
              section={sections[sectionKey]}
              currentSlug={currentSlug}
              onItemClick={onItemClick}
              readSections={readSections}
            />
          )
      )}
    </nav>
  );
};

export default Navigation;
