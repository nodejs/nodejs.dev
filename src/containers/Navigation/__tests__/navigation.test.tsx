import React from 'react';
import { render } from '@testing-library/react';
import Navigation from '..';

describe('Navigation component', (): void => {
  it('renders correctly', (): void => {
    const { container } = render(
      <Navigation
        currentSlug="intro"
        sections={{
          'Getting Started': [
            {
              slug: 'intro',
              title: 'Section 1',
              section: '...',
            },
          ],
          Versions: [
            {
              slug: 'version',
              title: 'Version 1',
              section: '...',
            },
          ],
        }}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
