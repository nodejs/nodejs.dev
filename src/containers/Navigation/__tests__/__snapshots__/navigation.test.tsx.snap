// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation component renders correctly 1`] = `
<div>
  <nav
    class="side-nav"
  >
    <button
      class="side-nav__open"
      type="button"
    >
      Menu
    </button>
    <ul
      class="side-nav__list"
    >
      <h2
        class="t-body2 side-nav__title"
      >
        <i
          class="material-icons"
        >
          offline_bolt
        </i>
        Getting Started
      </h2>
      <a
        class="t-body2 side-nav__item side-nav__item--active"
        href="/learn/intro"
        id="link-intro"
      >
        Section 1
      </a>
    </ul>
    <ul
      class="side-nav__list"
    >
      <h2
        class="t-body2 side-nav__title"
      >
        <i
          class="material-icons"
        >
          offline_bolt
        </i>
        Versions
      </h2>
      <a
        class="t-body2 side-nav__item"
        href="/learn/version"
        id="link-version"
      >
        Version 1
      </a>
    </ul>
  </nav>
</div>
`;
